pkg/services/ngalert/state/state_tracker.go

```diff
@@ -54,135 +56,158 @@ func NewStateTracker(logger log.Logger) *StateTracker {
 }
 
 //Used to ensure a clean cache on startup
 func (st *StateTracker) ResetCache() {
-	st.stateCache.mu.Lock()
-	defer st.stateCache.mu.Unlock()
-	st.stateCache.cacheMap = make(map[string]AlertState)
+	st.cache.mtxStates.Lock()
+	defer st.cache.mtxStates.Unlock()
+	st.cache.states = make(map[string]AlertState)
 }
 
-func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, results eval.Results) []AlertState {
+func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, results eval.Results, evaluationDuration time.Duration) []AlertState {
 	st.Log.Info("state tracker processing evaluation results", "uid", alertRule.UID, "resultCount", len(results))
-	var changedStates []AlertState
+	var states []AlertState
 	for _, result := range results {
-		s, _ := st.setNextState(alertRule, result)
-		changedStates = append(changedStates, s)
+		s := st.setNextState(alertRule, result, evaluationDuration)
+		states = append(states, s)
 	}
-	st.Log.Debug("returning changed states to scheduler", "count", len(changedStates))
-	return changedStates
+	st.Log.Debug("returning changed states to scheduler", "count", len(states))
+	return states
 }
 
 //TODO: When calculating if an alert should not be firing anymore, we should take three things into account:
 // 1. The re-send the delay if any, we don't want to send every firing alert every time, we should have a fixed delay across all alerts to avoid saturating the notification system
-// 2. The evaluation interval defined for this particular alert - we don't support that yet but will eventually allow you to define how often do you want this alert to be evaluted
-// 3. The base interval defined by the scheduler - in the case where #2 is not yet an option we can use the base interval at which every alert runs.
 //Set the current state based on evaluation results
-//return the state and a bool indicating whether a state transition occurred
-func (st *StateTracker) setNextState(alertRule *ngModels.AlertRule, result eval.Result) (AlertState, bool) {
-	currentState := st.getOrCreate(alertRule, result)
+func (st *StateTracker) setNextState(alertRule *ngModels.AlertRule, result eval.Result, evaluationDuration time.Duration) AlertState {
+	currentState := st.getOrCreate(alertRule, result, evaluationDuration)
 	st.Log.Debug("setting alert state", "uid", alertRule.UID)
 	switch {
 	case currentState.State == result.State:
 		st.Log.Debug("no state transition", "cacheId", currentState.CacheId, "state", currentState.State.String())
 		currentState.LastEvaluationTime = result.EvaluatedAt
+		currentState.EvaluationDuration = evaluationDuration
 		currentState.Results = append(currentState.Results, StateEvaluation{
 			EvaluationTime:  result.EvaluatedAt,
 			EvaluationState: result.State,
 		})
 		if currentState.State == eval.Alerting {
-			currentState.EndsAt = result.EvaluatedAt.Add(40 * time.Second)
+			currentState.EndsAt = result.EvaluatedAt.Add(alertRule.For * time.Second)
```

Comment:
Acquainting myself with this code, but is this is incorrect? The current state of an alert can end _any time_ it's evaluated again. The `for` duration of an alert rule determines how long that rule must be triggered for it to "fire", not how long it's considered "firing". Context around this: It's helpful for alerts to resolve themselves quickly after things get better.

Reply:
The logic here has changed in https://github.com/grafana/grafana/pull/33194 but it's still close to this. EndsAt is set by using the alertRule.For based on a conversation with, I believe @gotjosh but I don't have it at hand. As I understand things, EndsAt would come into play when no alerts have been sent to the alert manager for longer than EndsAt. Each evaluation that continues to be alerting sets the EndsAt further out and then anything that isn't normal/ok is sent to the notifier. 

If I'm wrong here then I would love to fix it and get it right.